File: 2022/6/day6.exs
defmodule Day6 do
  def main() do
    buf = IO.read(:line) |> String.trim() |> String.to_charlist()

    buf |> find_marker(4) |> IO.puts()
    buf |> find_marker(14) |> IO.puts()
  end

  defp find_marker(buf, len), do: find_marker(buf, len, 0)
  defp find_marker([first|rest], len, pos) do
    marker = Enum.take([first|rest], len)
    if length(marker) == MapSet.size(MapSet.new(marker)) do
      pos + len
    else
      find_marker(rest, len, pos + 1)
    end
  end
end

Day6.main()

File: 2022/6/day6.py
from sys import stdin

def find_marker(buf: str, marker_len: int) -> int:
    for i in range(len(buf)):
        marker = buf[i:i+marker_len]
        if len(marker) == len(set(marker)):
            return i + marker_len

    return -1

def main() -> None:
    buf = stdin.read().strip()

    print(find_marker(buf, 4))
    print(find_marker(buf, 14))

if __name__ == '__main__':
    main()
